src/svd/two-sided/svd.cpp

```diff
@@ -0,0 +1,132 @@
+    double off_norm = 0.0;  // frobenius norm of the off-diagonal elements of matrix
+
+    matrix_copy(Bmat, Amat);
+    matrix_identity(Umat);
+    matrix_identity(Vmat);
+    matrix_frobenius(Bmat, &norm, &off_norm);
+
+    while (off_norm >= tol * norm) {
+        for (size_t i = 0; i < n - 1; ++i) {
+            for (size_t j = i + 1; j < n; ++j) {
+                const double bii = B[n * i + i];
+                const double bij = B[n * i + j];
+                const double bji = B[n * j + i];
+                const double bjj = B[n * j + j];
+
+                struct svd_2x2_params cf = nsvd(bii, bij, bji, bjj);
+
+                // R_ij(c,s) * B where R_ij(c,s) is the Givens rotation matrix that acts on
+                // rows i and j during left multiplication with B
+                for (size_t k = 0; k < n; ++k) {
+                    double left = cf.c1 * B[n * i + k] - cf.s1 * B[n * j + k];
+                    double right = cf.s1 * cf.k * B[n * i + k] + cf.c1 * cf.k * B[n * j + k];
+                    B[n * i + k] = left;
+                    B[n * j + k] = right;
+                }
+
+                // B * R_ij(c',-s') where R_ij(c',-s') is the Givens rotation matrix that acts on
+                // columns i and j during right multiplication with B
+                for (size_t k = 0; k < n; ++k) {
+                    double left = cf.c2 * B[n * k + i] - cf.s2 * B[n * k + j];
+                    double right = cf.s2 * B[n * k + i] + cf.c2 * B[n * k + j];
+                    B[n * k + i] = left;
+                    B[n * k + j] = right;
+                }
+
+                // U * R_ij(c,s) where R_ij(c,s) is the Givens rotation matrix that acts on
+                // columns i and j during right multiplication with U
+                for (size_t k = 0; k < n; ++k) {
+                    double left = cf.c1 * U[n * k + i] - cf.s1 * U[n * k + j];
+                    double right = cf.s1 * cf.k * U[n * k + i] + cf.c1 * cf.k * U[n * k + j];
+                    U[n * k + i] = left;
+                    U[n * k + j] = right;
+                }
+
+                // V * R_ij(c',-s') where R_ij(c',-s') is the Givens rotation matrix that acts on
+                // columns i and j during right multiplication with V
+                for (size_t k = 0; k < n; ++k) {
+                    double left = cf.c2 * V[n * k + i] - cf.s2 * V[n * k + j];
+                    double right = cf.s2 * V[n * k + i] + cf.c2 * V[n * k + j];
+                    V[n * k + i] = left;
+                    V[n * k + j] = right;
+                }
+            }
+        }
+
+        matrix_off_frobenius(Bmat, &off_norm);
+        iter += 1;
+    }
+
+    for (size_t i = 0; i < n; ++i) {
+        s[i] = B[n * i + i];
+    }
+
+    return iter;
+}
+
+void matrix_frobenius(matrix_t m, double* norm, double* off_norm) {
+    const size_t M = m.rows;
+    const size_t N = m.cols;
+    double* data = m.ptr;
+
+    for (size_t i = 0; i < M; ++i) {
```

Comment:
I think `off_norm` should be initialized to 0.0 before the loop in this function too, similar to `matrix_off_frobenius`

Reply:
You are right, both ```norm``` and ```off_norm``` should be initialized to 0. I updated the code.